components/eamxx/src/diagnostics/pbl_entrainment_budget.cpp

```diff
@@ -0,0 +1,332 @@
+            tm_v(icol, jlev), qc_v(icol, jlev));
+      });
+}
+
+void PBLEntrainmentBudget::init_timestep(const util::TimeStamp &start_of_step) {
+  m_start_t = start_of_step;
+
+  const auto &tm_v = get_field_in("T_mid").get_view<Real **>();
+  const auto &pm_v = get_field_in("p_mid").get_view<Real **>();
+  const auto &qv_v = get_field_in("qv").get_view<Real **>();
+  const auto &qc_v = get_field_in("qc").get_view<Real **>();
+
+  const auto &m_prev_qt_v = m_prev_qt.get_view<Real **>();
+  const auto &m_prev_tl_v = m_prev_tl.get_view<Real **>();
+
+  calc_tl_qt(tm_v, pm_v, qv_v, qc_v, m_prev_tl_v, m_prev_qt_v);
+}
+
+void PBLEntrainmentBudget::compute_diagnostic_impl() {
+  using PC  = scream::physics::Constants<Real>;
+  using KT  = KokkosTypes<DefaultDevice>;
+  using MT  = typename KT::MemberType;
+  using ESU = ekat::ExeSpaceUtils<typename KT::ExeSpace>;
+
+  constexpr Real g = PC::gravit;
+  Real fill_value  = constants::DefaultFillValue<Real>().value;
+
+  // Before doing anything, subview the out field for each variable
+  auto out = m_diagnostic_output.get_view<Real **>();
+
+  auto o_pm_hplus = ekat::subview_1(out, m_index_map["p+"]);
+  auto o_tl_hplus = ekat::subview_1(out, m_index_map["tl+"]);
+  auto o_tl_caret = ekat::subview_1(out, m_index_map["tl^"]);
+  auto o_tl_ttend = ekat::subview_1(out, m_index_map["tl_ttend"]);
+  auto o_qt_hplus = ekat::subview_1(out, m_index_map["qt+"]);
+  auto o_qt_caret = ekat::subview_1(out, m_index_map["qt^"]);
+  auto o_qt_ttend = ekat::subview_1(out, m_index_map["qt_ttend"]);
+  auto o_df_inpbl = ekat::subview_1(out, m_index_map["dF"]);
+
+  // Get the input views
+  const auto &qc_v = get_field_in("qc").get_view<Real **>();
+  const auto &qv_v = get_field_in("qv").get_view<Real **>();
+  const auto &tm_v = get_field_in("T_mid").get_view<Real **>();
+  const auto &pm_v = get_field_in("p_mid").get_view<Real **>();
+  const auto &pd_v = get_field_in("pseudo_density").get_view<Real **>();
+  const auto &sd_v = get_field_in("SW_flux_dn").get_view<Real **>();
+  const auto &su_v = get_field_in("SW_flux_dn").get_view<Real **>();
+  const auto &ld_v = get_field_in("LW_flux_dn").get_view<Real **>();
+  const auto &lu_v = get_field_in("LW_flux_up").get_view<Real **>();
+
+  // tracked stuff
+  const auto &prev_qtot_v = m_prev_qt.get_view<Real **>();
+  const auto &prev_tliq_v = m_prev_tl.get_view<Real **>();
+
+  view_2d qt_v("qt_v", m_ncols, m_nlevs);
+  view_2d tl_v("tl_v", m_ncols, m_nlevs);
+  calc_tl_qt(tm_v, pm_v, qv_v, qc_v, tl_v, qt_v);
+
+  const auto &curr_ts =
+      get_field_in("qc").get_header().get_tracking().get_time_stamp();
+
+  auto dt = curr_ts - m_start_t;
+
+  const int num_levs  = m_nlevs;
+  const int pblinvalg = m_pblinvalg;
+  const auto policy   = ESU::get_default_team_policy(m_ncols, m_nlevs);
+
+  constexpr int wsms = 1;
+  using WSMgr        = ekat::WorkspaceManager<Real, DefaultDevice>;
+  WSMgr wsm(num_levs, wsms, policy);
```

Comment:
This allocates memory at every evaluation. You could move the creation of the wsm to `initialize_impl`.